### PR TITLE
Unsafe paste. (#45)

### PR DESCRIPTION
**Pull request by LightSunrider**

### CODE REVIEW

**Comment by danrabbit** [comment]
Can you explain what the intended change here is? Is there an issue that this is intended to resolve?

**Comment by LightSunrider** [comment]
Unfortunately my knowledge is not very broad and I can make a mistake (if I have not already done that :) )

In the terminal widget (vte, gtk or the other, I'm not sure) have their own code to insert text from the clipboard by pressing the middle mouse button. Because of this, you can insert non-secure commands without warning. This can be fixed by own 'middle mouse button pressing' signal handler for paste.

I tested better and realized that my solution does not work on the restored tabs. I think that the beautiful solution will require a little refactoring (what I'm afraid and decided not to do so as not to spoil the existing code).

**Comment by LightSunrider** [comment]
I have carefully read code of the widgets. Gtk.Container automatically destroys the child widget. To avoid this, the child widget should be removed.

My solution should fix #45, #88, #94, and links in restored tabs (they were not clickable)

File: src/PantheonTerminalWindow.vala [hunk 1]
```diff
@@ -658,6 +670,8 @@ namespace PantheonTerminal {
             restorable_terminals.insert (term.terminal_id, term);
             tab.restore_data = term.terminal_id;
 
+            ((Gtk.Container)tab.page).remove(term);
```
Comment (donadigo) [inline src/PantheonTerminalWindow.vala]: Wouldn't `term.destroy ()` work here?

File: src/PantheonTerminalWindow.vala [hunk 2]
```diff
@@ -613,6 +613,18 @@ namespace PantheonTerminal {
...
 
+            t.button_press_event.connect ((e) => {
+                if (e.button == Gdk.BUTTON_MIDDLE) {
+                    if (this.search_toolbar.search_entry.has_focus) {
+                        return false;
+                    } else if (clipboard.wait_is_text_available ()) {
+                        action_paste ();
+                        return true;
+                    }
+                }
```
Comment (donadigo) [inline src/PantheonTerminalWindow.vala]: Add a whitespace here.

File: src/PantheonTerminalWindow.vala (hunk 1 above)
Comment (LightSunrider) [inline src/PantheonTerminalWindow.vala]: Sorry, I don't know what you meant. If after this line add `term.destroy ()`, then all term signals will be disabled

Comment (donadigo) [inline src/PantheonTerminalWindow.vala]: @LightSunrider ah yes, you're right here.

**Comment by LightSunrider** [comment]
I should revert 9a1ceb83ff1bd3cba2e28cfefe659914b66bb0f2 in this branch and create new pull request with this commit in new branch?

**Comment by donadigo** [comment]
@LightSunrider yeah it would be ideal.

**Comment by LightSunrider** [comment]
Thanks, I hope I did it right

**Comment by LightSunrider** [comment]
Can you review my changes? They are incorrect or incomplete?

**Comment by donadigo** [comment]
Ups, sorry, I forgot I reviewed this before. I will review it today / tomorrow. I removed the Incomplete label.

File: src/Widgets/TerminalWidget.vala [hunk 3]
```diff
@@ -130,6 +130,9 @@ namespace PantheonTerminal {
 
                     return true;
                 }
+                else if (event.button == Gdk.BUTTON_MIDDLE) {
```
Comment (donadigo) [inline src/Widgets/TerminalWidget.vala]: Make this `else if` on the same line as the above `}`:

```
...
} else if (...) {

}
```

File: src/Widgets/TerminalWidget.vala [hunk 4]
```diff
@@ -130,6 +130,9 @@ namespace PantheonTerminal {
...
+                    return window.handle_paste_event();
```
Comment (donadigo) [inline src/Widgets/TerminalWidget.vala]: Add a space between the `handle_paste_event` and `()`.

File: src/PantheonTerminalWindow.vala [hunk 5]
```diff
@@ -319,6 +319,17 @@ namespace PantheonTerminal {
             });
         }
 
+        public bool handle_paste_event () {
```
Comment (donadigo) [inline src/PantheonTerminalWindow.vala]: Could this be called also here?
https://github.com/LightSunrider/terminal/blob/97fbeeb5949e907b988fc0f20534a1fd41d62180/src/PantheonTerminalWindow.vala#L309

This is the same code as in this method.

**Comment by LightSunrider** [comment]
Thank you for your comments. I fixed it.

File: src/PantheonTerminalWindow.vala [hunk 6]
```diff
@@ -306,19 +306,25 @@ namespace PantheonTerminal {
...
+                        handle_paste_event ();
                     }
                 }
 
                 return false;
             });
         }
 
+        public bool handle_paste_event () {
+            if (this.search_toolbar.search_entry.has_focus) {
```
Comment (jeremypw) [inline src/PantheonTerminalWindow.vala]: Unnecessary use of ****this****.  There is no ambiguity and rest of this class does not use this form for `searchbar`.

**Comment by LightSunrider** [comment]
Fixed. All right now?

**Pull request merged**